### PR TITLE
Chemisty recipe macros

### PR DESCRIPTION
**Pull request by ghost**
Ripped from @Iamgoofball's PR

:cl: More Robust Than You
add: You can now save chemistry recipe macros
/:cl:

### CODE REVIEW

**Comment by kevinz000** [comment]
welp there goes the only thing stopping me from powergaming chemistry.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -179,6 +185,31 @@
 				beaker = null
 				cut_overlays()
 				. = TRUE
+		if("dispense_recipe")
+			var/recipe_to_use = params["recipe"]
+			var/list/chemicals_to_dispense = process_recipe_list(recipe_to_use)
+			for(var/r_id in chemicals_to_dispense) // i suppose you could edit the list locally before passing it
+				if(beaker && dispensable_reagents.Find(r_id)) // but since we verify we have the reagent, it'll be fine
+					var/amt = chemicals_to_dispense[r_id]
+					beaker.reagents.add_reagent(r_id, amt)
```
Comment (kevinz000) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: lol you're not even using cell energy this is even better than i thought!

**Comment by praisenarsie** [comment]
how does this work exactly

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (MoreRobustThanYou) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: thanks for pointing this out, now i'll fix it

**Comment by MoreRobustThanYou** [comment]
@praisenarsie add recipe in this format

```chemical_id=amount;chemical_id=amount```. Only chemicals that the machine can dispense will work. This also works with drink/booze machines.

**Comment by praisenarsie** [comment]
so essentially like goon?

like goon where it only saves the recipe on that machine?
like goon where it only saves that recipe for that one round only?

**Comment by MoreRobustThanYou** [comment]
yes and yes

**Comment by praisenarsie** [comment]
ok morty did a good

👍

**Comment by praisenarsie** [comment]
actually wait we need Need Pictures

**Comment by MoreRobustThanYou** [comment]
sure

**Comment by ShiggyDiggyDo** [comment]
If someone writes down macros for chemicals on the wiki chemists may actually fill up the fridge with medicine for one

**Comment by praisenarsie** [comment]
@ShiggyDiggyDo you must have really shitty rounds

**Comment by praisenarsie** [comment]
@MoreRobustThanYou woah slowdown their partner, this makes it so you can input a custom amount of reageants so you can do 33u even though that's not possible from the given buttons.

make it so you can choose a custom amount for dispensing reagents

**Comment by MoreRobustThanYou** [comment]
@praisenarsie out of scope

**Comment by praisenarsie** [comment]
out of scope my ass, you're blatantly making an oversight/consistency issue in a pr

**Comment by 81Denton** [comment]
@ShiggyDiggyDo Check out the wiki's "Chemical_recipes" page

**Comment by Buggy123** [comment]
So this is basically the chemistry redesign Goof did where he added unlimited energy AND recipe programming a-la goon, except this time it's just the recipe programming?

:+1:

**Comment by coiax** [comment]
This is fine, provided we eventually remove unlimited reagent dispensal.

**Comment by Buggy123** [comment]
We already don't have unlimited reagent dispensal. Unless you mean remove regenerating energy, fuck no.

**Comment by FantasticFwoosh** [comment]
Plant growing is too slow and ineffectual at the beginning of the round to not have atleast a ineffectual infinite chemistry source @Buggy123 if you wanted to get some raw chemicals, making more morphine from reishi takes ages without plant optimisation and preparation in growing etc.

**Comment by Bawhoppen** [comment]
Chemistry is already pretty broken, this will just break it more under the guise of being "QoL"
(Or it might make it more obvious how it's broken, leading to someone actually fixing it eventually)

**Comment by Buggy123** [comment]
Holy shit what, Coiax you want to REMOVE reagent dispensing?

If that shit even remotely goes through I'll just make a PR removing medbay.

**Comment by Bawhoppen** [comment]
Reagent dispensing is incredibly silly, and definitely somewhat broken. We need a better system for gathering ingredient reagents that comes from different sources including mining, botany, cargo, and certainly science.

**Comment by TribeOfBeavers** [comment]
A lot of the higher tier chems are balanced around the fact that they're annoying and fiddly to make. Being able to output custom amounts by copy/pasting a recipe will make it significantly easier to get them and especially easier to mass produce them.

This also removes any skill celing from chemistry. Currently more experienced chemists can make chems faster and in larger amounts compared to newer ones. This change makes it most efficient to simply copy and paste the recipe, as it allows for exact amounts of something to be made faster than by hand so there is no reason to ever do it.

QoL changes are rarely a good thing in this game, to be honest.

**Comment by Iamgoofball** [comment]
>A lot of the higher tier chems are balanced around the fact that they're annoying and fiddly to make. Being able to output custom amounts by copy/pasting a recipe will make it significantly easier to get them and especially easier to mass produce them.

they're not, actually

they're balanced around having downsides and side effects

# oh wait you idiots readded trekchems which have no downsides or side effects LMAO

**Comment by Iamgoofball** [comment]
@Bawhoppen go home with your realism arguments, nobody of note actually cares about them

**Comment by AdamElTablawy** [comment]
Sulfonal-powerful poison, rarely made because it takes a bit of finagling.

With change, easy to massproduce.

With this change you may as well just change it so that all the buttons on the chemmaster are the full list of mixable chems rather than the components to mix them.

**Comment by 81Denton** [comment]
>A lot of the higher tier chems are balanced around the fact that they're annoying and fiddly to make. Being able to output custom amounts by copy/pasting a recipe will make it significantly easier to get them and especially easier to mass produce them.

I have to disagree with that - even Pentetic Acid, Skewium, Corgium, etc. are very easy to make if you sit down for five minutes and write down the steps and amounts needed.
For some reason, most chemists can't be assed to do that though. Lack of patience? That's why you only see them make babby chems and maybe Synthflesh at best.

**Comment by 81Denton** [comment]
>Sulfonal-powerful poison, rarely made because it takes a bit of finagling.
>With change, easy to massproduce.

All the stuff that makes chemist traitors strong (pyro and stun chems) is already very easy to mix.
The poisons you can make are pretty awful compared to what you get from the poison kit (Initropidil, Curare, Pancuronium) or Botany (Coniine, Amanitin).

**Comment by Buggy123** [comment]
I've started playing chemist, and honestly, its not that hard. Maybe its just because my other main job is shaft miner, aka powergaming: the job, but a little practice and memorization and you've got it, the rest of chemistry is just experience and knowing what chems to use where at the best ratios to get the desired result, the sort of thing that exists with any job.

Skewnium is a bit annoying but I've gotten the hang of mixing it, takes me 5 ish minutes and I could cut it down a lot with more practice. Sulfonal looks piss easy TBH. Off the top of my head, just checking the wiki for sulfonal itself, its... 1-1-1 acetone dimethyl sulfur. So then, lets see if i get this right:
5 weld 5 hydr 5 carb>15 oil+.... 15 weld 15 o2? gets 45 acet, dump 15.
5 weld 5 hydr 5 carb>15 oil+15 ethanol>30 dimethyl.
30 sulfur straight from the machine, bam, 'finagling' done, 90 sulfonal.


Chemistry is not hard. I don't think this changes much, except for lowering the skill gap, but it hardly removes it completely and I don't think that's a particularly negative change.

**Comment by CPTANT** [comment]
@Iamgoofball 
The fuck man, its Trekchems that DO have overdoses, its your healing patches that have NO OVERDOSES OR OTHER NEGATIVE EFFECTS.

**Comment by CPTANT** [comment]
Anyway this is a serious balance change.

The machine will now just have a button that goes print healing patch *1000

**Comment by MoreRobustThanYou** [comment]
Chemists almost NEVER leave out healing chems.

**Comment by PKPenguin321** [comment]
Goof
Trekchems are objectively worse than goofchems for healing
How many times do we have to teach you this lesson old man

**Comment by FantasticFwoosh** [comment]
@MoreRobustThanYou i've seen this happen on other servers, somehow chemists actively choose to not distribute chems to the public for RP reasons (durrhurr you need a prescription) or to deny antagonists opportunities to heal and therefore allow powergaming to attack weakened antagonists. 

- Because usually grabbing medical supplies is high on a wizard's personal list of to-do tasks.

ITT you are all overreacting because you've forgotten that some chemicals require to be heated/chilled/stabilised before undergoing further chemical changes, if you simply widen the amount of chemicals that need these processes, the narrower the scope of macro's will be to be actually QOL.

**Comment by MoreRobustThanYou** [comment]
a new era in chemistry has begun

**Comment by davethwave** [comment]
oh my

**Comment by Bawhoppen** [comment]
Biggest mistake of the week

**Comment by ACCount12** [comment]
Best PR of the year.

**Comment by Bawhoppen** [comment]
A year is a long time, are you sure?

**Comment by 81Denton** [comment]
The most powerful griff chems are easy to mix and the complicated stuff requires either multiple beakers or removing/separating ingredients. This will barely change anything for chem powergamers.

**Comment by Iamgoofball** [comment]
Yeah all this does is prevent carpal tunnel syndrome

**Comment by Buggy123** [comment]
Honestly haven't even touched this feature yet despite mostly playing as chemist since it's been merged.

**Comment by Robustin** [comment]
Because its virtually unusable. I read the code and I still can't get it to work.

**Comment by MoreRobustThanYou** [comment]
@Robustin make sure you follow the exact format

**Pull request merged**